src/Symfony/Component/Security/Http/Attribute/IsGranted.php

```diff
@@ -24,18 +26,37 @@
 final class IsGranted
 {
     /**
-     * @param string|Expression            $attribute     The attribute that will be checked against a given authentication token and optional subject
-     * @param array|string|Expression|null $subject       An optional subject - e.g. the current object being voted on
-     * @param string|null                  $message       A custom message when access is not granted
-     * @param int|null                     $statusCode    If set, will throw HttpKernel's HttpException with the given $statusCode; if null, Security\Core's AccessDeniedException will be used
-     * @param int|null                     $exceptionCode If set, will add the exception code to thrown exception
+     * @param string|Expression|null                                                           $attribute     The attribute that will be checked against a given authentication token and optional subject, or a callable that will be called to determine access
+     * @param array|string|Expression|\Closure(Request $request, array $arguments): mixed|null $subject       An optional subject - e.g. the current object being voted on
+     * @param string|null                                                                      $message       A custom message when access is not granted
+     * @param int|null                                                                         $statusCode    If set, will throw HttpKernel's HttpException with the given $statusCode; if null, Security\Core's AccessDeniedException will be used
+     * @param int|null                                                                         $exceptionCode If set, will add the exception code to thrown exception
+     * @param \Closure(AuthorizationCheckerInterface, mixed $subject): bool|null               $callable      A callable that will be called to determine access
      */
     public function __construct(
-        public string|Expression $attribute,
-        public array|string|Expression|null $subject = null,
+        public string|Expression|null $attribute = null,
+        public array|string|Expression|\Closure|null $subject = null,
         public ?string $message = null,
         public ?int $statusCode = null,
         public ?int $exceptionCode = null,
+        public ?\Closure $callable = null,
     ) {
+        if (\PHP_VERSION_ID < 80500) {
+            if (null === $attribute) {
+                throw new \LogicException('The "attribute" argument is only optional starting from PHP 8.5.');
+            }
+
+            if (null !== $callable) {
+                throw new \LogicException('The "callable" argument is usable only starting from PHP 8.5.');
+            }
+
+            if (\is_callable($subject)) {
+                throw new \LogicException('The "subject" argument can be a callable only starting from PHP 8.5.');
+            }
```

Comment:
By default, is_callable returns true for every array, strings, etc.. that "looks" callable. 

https://www.php.net/manual/fr/function.is-callable.php

Reply:
Oops nope... but still a big problem here :) 


Many string will be callable

Reply:
What's an concrete example of what you have in mind? A subject called "trim" for instance?

Reply:
In one of my older projects I use string subjects as short names for entity classes. So there is a MailVoter, which votes on `App\Entity\Mail` objects, but also on the string subject `mail` as short name for the Mail entity. Voting on (short) class names is used for general permissions regarding these objects not depending on concrecte objects, e.g. `is_granted('create', 'mail')`.

I don't know if I would build it that way again, but it is a concrete example of a string subject (`'mail'`), [which is `callable`](https://www.php.net/manual/en/function.mail.php).